Update instance segmentation

DmitriySidnev opened the pull request:
These are new instance segmentation models.

eaidova commented:
Jenkins ok to test

eaidova commented on models/intel/instance-segmentation-security-0091/accuracy-check.yml:
```diff
@@ -21,11 +17,12 @@ models:
             dst_height: 800
             pad_type: right_bottom
-            use_numpy: True
+            use_numpy: True 
```
```suggestion
            use_numpy: True
```

eaidova commented on tools/accuracy_checker/accuracy_checker/adapters/mask_rcnn.py:
```diff
@@ -240,7 +244,10 @@ def _process_masks_pytorch(self, boxes, raw_masks, identifiers, original_image_s
             per_obj_raw_masks = []
             for cls, raw_mask in zip(classes, raw_masks):
-                per_obj_raw_masks.append(raw_mask[cls, ...])
+                if self.scores_out:
+                    per_obj_raw_masks.append(raw_mask[cls, ...])
+                else:
+                    per_obj_raw_masks.append(raw_mask)
```
fast fix for linter issue
```suggestion
                    per_obj_raw_masks.append(raw_mask[cls, ...] if self.scores_out else raw_mask)
```

eaidova commented:
@DmitriySidnev please add new models to index.md

druzhkov-paul commented on models/intel/instance-segmentation-security-0002/description/instance-segmentation-security-0002.md:
```diff
@@ -0,0 +1,45 @@
+|---------------------------------|-------------------------------------------|
+| MS COCO val2017 box AP          | 40.8%                                     |
+| MS COCO val2017 mask AP         | 36.9%                                     |
+| Max objects to detect           | 1000                                      |
```
This 1000 does not match with shape of network outputs described below.

DmitriySidnev replied:
Fixed to the output shape.

eizamaliev commented on models/intel/instance-segmentation-security-0002/description/instance-segmentation-security-0002.md:
```diff
@@ -0,0 +1,45 @@
+
+## Outputs
+
+1.	name: `labels`, shape: [100, ] - Contiguous integer class ID for every
```
May be `[100,1]` or `[100]`?

DmitriySidnev replied:
Fixed to `[100]`.

vladimir-dudnik commented on models/intel/instance-segmentation-security-0002/description/instance-segmentation-security-0002.md:
```diff
@@ -0,0 +1,45 @@
+Average Precision (AP) is defined and measured according to standard
+[MS COCO evaluation procedure](http://cocodataset.org/#detection-eval).
+
+## Performance
```
we are removing empty sections from documentation (see [PR-1749](https://github.com/openvinotoolkit/open_model_zoo/pull/1749)), please do this as well

DmitriySidnev replied:
Fixed.

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
please add soft link to AC config from old configs location

vladimir-dudnik reviewed (changes requested):
do not merge until accuracy validation release criteria met

DmitriySidnev commented:
@vladimir-dudnik, @eaidova, should I do something else except accuracy-checker results? CI failed for Demos because demos ran with old models that are not compatible with new ones and Documentation because in `index.md` are not present old models now.

vladimir-dudnik commented:
@DmitriySidnev we will return to this PR once new models merged. This will not go for 2021.2 release.

DmitriySidnev commented:
@vladimir-dudnik, I have updated metrics. For two models 0002 and 0091 that can't be preprocessed as in Pytorch I added metrics for both options as it was early for old models. For other models also were slightly reduced metrics. Please take a look.

vladimir-dudnik reviewed (approved):
model now fits release criteria

vladimir-dudnik commented:
@DmitriySidnev in docs there are link to replaced models:
17:57:51  [✖] ./instance-segmentation-security-0904/description/instance-segmentation-security-0904.md
17:57:51  [✖] ./instance-segmentation-security-0912/description/instance-segmentation-security-0912.md

vladimir-dudnik commented:
@DmitriySidnev now the issue is that:
13:10:23  instance-segmentation-security-0050 not in index.md file

eaidova commented on tools/accuracy_checker/accuracy_checker/adapters/mask_rcnn.py:
```diff
@@ -175,8 +175,12 @@ def _process_pytorch_outputs(self, raw_outputs, identifiers, frame_meta):
+            classes = classes[valid_detections_mask]
+        else:
+            valid_detections_mask = np.sum(boxes, axis=1) > 0
+            classes = classes[valid_detections_mask] + 1
```
not sure that it is necessary, we prefer to align classes in annotation with model labels, there is option which allow convert dataset without background label

DmitriySidnev replied:
Fixed.

vladimir-dudnik commented:
14:18:50  demos/instance_segmentation_demo/python/instance_segmentation_demo/model_utils.py:239: trailing whitespace.
14:18:50  +    'mask_rcnn_segmentoly': ModelAttributes(('boxes', 'scores', 'classes', 'raw_masks'), 
14:18:50  running yamllint...
14:18:56  running flake8...
14:19:04  ./demos/whiteboard_inpainting_demo/python/utils/network_wrappers.py:37:16: E713 test for membership should be 'not in'
14:19:04  ./demos/multi_camera_multi_target_tracking/python/utils/network_wrappers.py:149:16: E713 test for membership should be 'not in'

IRDonch commented on models/intel/index.md:
```diff
@@ -140,6 +140,11 @@ instance instance segmentation model outputs pixel-wise masks for all instances.
+| [instance-segmentation-security-1040](./instance-segmentation-security-1040/description/instance-segmentation-security-1040.md)                | 29.334               | 13.5673    |
+| [instance-segmentation-security-0228](./instance-segmentation-security-0228/description/instance-segmentation-security-0228.md)                | 147.2352             | 49.8328    |
+| [instance-segmentation-security-0002](./instance-segmentation-security-0002/description/instance-segmentation-security-0002.md)                | 423.0842             | 48.3732    |
+| [instance-segmentation-security-0091](./instance-segmentation-security-0091/description/instance-segmentation-security-0091.md)                | 828.6324             | 101.236    |
```
Please list the models in alphabetic (well, numeric, in this case) order.

DmitriySidnev replied:
Fixed.

IRDonch commented on models/intel/instance-segmentation-security-0002/description/instance-segmentation-security-0002.md:
```diff
@@ -0,0 +1,45 @@
+| Source framework                                                    | PyTorch\*                                 |
+
+Average Precision (AP) is defined and measured according to standard
+[MS COCO evaluation procedure](http://cocodataset.org/#detection-eval).
```
```suggestion
[MS COCO evaluation procedure](https://cocodataset.org/#detection-eval).
```

DmitriySidnev replied:
Fixed documentation for all new models.

Wovchena commented:
Please, extend tests with new models: https://github.com/openvinotoolkit/open_model_zoo/blob/29a8c8cef1cc60496ee04cbf05f8b8974893d238/demos/tests/cases.py#L481

DmitriySidnev commented on demos/instance_segmentation_demo/python/instance_segmentation_demo/model_utils.py:
```diff
@@ -56,14 +56,19 @@ def segm_postprocess(box, raw_cls_mask, im_h, im_w):
         outputs, scale_x, scale_y, frame_height, frame_width, input_height, input_width, conf_threshold
 ):
-    boxes = outputs['boxes']
+    segmentoly_postrpocess = 'raw_masks' in outputs
```
Fixed.

DmitriySidnev commented on demos/multi_camera_multi_target_tracking/python/utils/network_wrappers.py:
```diff
@@ -133,15 +133,21 @@ def __init__(self, ie, model_path, trg_classes, conf=.6,
-        required_input_keys = [{'im_info', 'im_data'}, {'im_data', 'im_info'}]
-        current_input_keys = self.net.input_info.keys()
+        required_input_keys = [{'image'}, {'im_info', 'im_data'}, {'im_data', 'im_info'}]
+        current_input_keys = set(self.net.inputs_info.keys())
```
Yes, removed.

IRDonch replied:
There's also similar code in `demos/whiteboard_inpainting_demo/python/utils/network_wrappers.py`.

DmitriySidnev replied:
Fixed.

DmitriySidnev commented on demos/whiteboard_inpainting_demo/python/utils/network_wrappers.py:
```diff
@@ -22,46 +22,65 @@ class MaskRCNN(IEModel):
+        self.n, self.c, self.h, self.w = self.inputs_info[input_name].input_data.shape
+
+    def segmentoly_type(self):
+        return 'image' not in set(self.inputs_info.keys())
```
Fixed.

DmitriySidnev commented on demos/whiteboard_inpainting_demo/python/utils/network_wrappers.py:
```diff
@@ -22,46 +22,65 @@ class MaskRCNN(IEModel):
-        im_info=np.array([processed_image.shape[1], processed_image.shape[2], 1.0], dtype='float32')
+        im_info = np.array([processed_image.shape[1], processed_image.shape[2], 1.0], dtype='float32')
+        if not self.segmentoly_type():
+            im_info = None
```
Fixed.

DmitriySidnev commented on demos/whiteboard_inpainting_demo/python/utils/network_wrappers.py:
```diff
@@ -22,46 +22,65 @@ class MaskRCNN(IEModel):
+            classes = output['classes'][valid_detections_mask]
+            boxes = output['boxes'][valid_detections_mask]
+            scores = output['scores'][valid_detections_mask]
+            masks = output['raw_masks'][valid_detections_mask]
```
Reimplemented method `segmentoly_type()` with checking all inputs and outputs.

IRDonch replied:
It's not quite sufficient, because your `__init__` method still admits models with segmentoly inputs and non-segmentoly outputs, and vice versa. Such models will be treated as non-segmentoly, and either input or output processing will fail.

DmitriySidnev replied:
Reimplemented.

DmitriySidnev commented:
```
15:05:56  + ./venv/bin/python ./src/demos/tests/run_tests.py --demo-build-dir=build/intel64/Release --test-data-dir=/mnt/datasets --downloader-cache-dir=/mnt/md-cache --mo=/home/jenkins/workspace/omz/precommit/demos_ubuntu18/openvino/deployment_tools/model_optimizer/mo.py --devices=CPU
15:05:57  Retrieving models...
15:05:59  No matching models: "instance-segmentation-security-0228"
```
Is this error caused by me?

IRDonch commented on demos/multi_camera_multi_target_tracking_demo/python/utils/network_wrappers.py:
```diff
@@ -133,15 +133,24 @@ def __init__(self, ie, model_path, trg_classes, conf=.6,
 
+    def check_segmentoly_type(self):
+        required_inputs = [{'im_info', 'im_data'}, {'im_data', 'im_info'}]
+        requiredd_outputs = {'boxes', 'scores', 'classes', 'raw_masks'}
```
```suggestion
        required_outputs = {'boxes', 'scores', 'classes', 'raw_masks'}
```

DmitriySidnev replied:
Thanks, fixed.

IRDonch commented:
> Is this error caused by me?

The new test cases can't run, because the new models aren't published yet. Comment them out, and we'll reenable them once we publish the models.

Also, note that the style check in the trigger build failed. Please fix the issues that were found.

vladimir-dudnik commented:
Jenkins please retry a build

Wovchena commented on demos/multi_camera_multi_target_tracking_demo/python/utils/network_wrappers.py:
```diff
@@ -133,15 +133,29 @@ def __init__(self, ie, model_path, trg_classes, conf=.6,
+        required_output_keys = {'boxes', 'labels', 'masks'}
 
-        self.n, self.c, self.h, self.w = self.net.input_info['im_data'].input_data.shape
+        required_input_keys_segmentoly = [{'im_info', 'im_data'}, {'im_data', 'im_info'}]
```
These two sets are equal. The same is for `check_segmentoly_type()` and whiteboard_inpainting_demo.

DmitriySidnev replied:
Fixed.

Wovchena reviewed (changes requested):
`py ..\instance_segmentation_demo\python\instance_segmentation_demo.py -m C:\Users\vzlobin\Downloads\instance-segmentation-security-1039.xml --labels C:\Users\vzlobin\a\z\data\dataset_classes\coco.txt -i 0`
failed with
`While validating node 'v6::ExperimentalDetectronPriorGridGenerator ExperimentalDetectronPriorGridGenerator_1766 (1069/Output_0/Data__const[0]:f16{3,4}, Conv_507[0]:f16{?,3,120,120}, 1070/Output_0/Data__const[0]:f16{1,1,1,1}) -> ()' with friendly_name 'ExperimentalDetectronPriorGridGenerator_1766':
The first dimension of both 'feature_map' and 'im_data' must match. Feature_map: ?; Im_data: 1`
I haven't aligned versions of MO and IE used yet, but may be you have a better guess why it failed.

DmitriySidnev commented:
`I haven't aligned versions of MO and IE used yet, but may be you have a better guess why it failed.`
@Wovchena, I have the only one suggestion - use IE with a model obtained via consistent version of MO.

Wovchena commented:
It still fails for me. I took IE and MO from 80781ff359f0b7bfd657cd3a78911c4ec108931b commit. Here is what I do
`v` folder stands for cloned https://github.com/openvinotoolkit/openvino, `z` is for https://github.com/openvinotoolkit/open_model_zoo
```PS1
cd a/v/build
cmake -G "Visual Studio 16 2019" -DENABLE_CPPLINT=OFF -DENABLE_CLDNN=OFF -DENABLE_HDDL=OFF -DENABLE_MYRIAD=OFF -DENABLE_TESTS=OFF -DENABLE_BEH_TESTS=OFF -DENABLE_FUNCTIONAL_TESTS=OFF -DENABLE_DLIA=OFF -DENABLE_PRIVATE_MODELS=OFF -DVERBOSE_BUILD=OFF -DENABLE_PERFORMANCE_TESTS=OFF -DENABLE_INTEGRATION_TESTS=OFF -DENABLE_PLUGIN_RPATH=ON -DENABLE_OPENVX_CVE=OFF -DOpenCV_DIR="C:\Program Files (x86)\IntelSWTools\openvino_2021\opencv\cmake" -DENABLE_ICV_TESTS=OFF -DENABLE_MODELS=OFF -DENABLE_OBJECT_DETECTION_TESTS=OFF -DENABLE_SEGMENTATION_TESTS=OFF -DENABLE_VALIDATION_SET=OFF -DENABLE_VPU=OFF -DENABLE_GNA=OFF -DENABLE_PROFILING_ITT=OFF -DTHREADING=TBB -DCMAKE_BUILD_TYPE=Release -DENABLE_PYTHON=ON .. ; cmake --build . -j 4 --config Release
cd ../../z/demos/build
py C:\Users\vzlobin\a\v\model-optimizer\mo_onnx.py --input_model C:\Users\vzlobin\Downloads\model.onnx --reverse_input_channels --mean_values "[ 103.53, 116.28, 123.675 ]" --output boxes,labels,masks --input_shape "[ 1, 3, 480, 480 ]"
$env:PATH="C:\Users\vzlobin\a\v\bin\intel64\Release;C:\Program Files (x86)\IntelSWTools\openvino_2021\opencv\bin;C:\Users\vzlobin\a\v\inference-engine\temp\tbb\bin;$env:PATH"; $env:PYTHONPATH="C:\Users\vzlobin\a\z\demos\build\intel64\Release;C:\Users\vzlobin\a\v\bin\intel64\Release\python_api\python3.7;C:\Program Files (x86)\IntelSWTools\openvino_2021\python\python3;$env:PYTHONPATH"
py ..\instance_segmentation_demo\python\instance_segmentation_demo.py -m C:\Users\vzlobin\a\z\demos\build\.\model.xml --labels C:\Users\vzlobin\a\z\data\dataset_classes\coco.txt -i 0
```
```
[ INFO ] Creating Inference Engine...
[ INFO ] Loading network
Traceback (most recent call last):
  File "..\instance_segmentation_demo\python\instance_segmentation_demo.py", line 239, in <module>
    sys.exit(main() or 0)
  File "..\instance_segmentation_demo\python\instance_segmentation_demo.py", line 110, in main
    net = ie.read_network(args.model, args.model.with_suffix('.bin'))
  File "ie_api.pyx", line 266, in openvino.inference_engine.ie_api.IECore.read_network
  File "ie_api.pyx", line 288, in openvino.inference_engine.ie_api.IECore.read_network
RuntimeError: Check 'num_batches_featmap.same_scheme(num_batches_im_data)' failed at C:\Users\vzlobin\a\v\ngraph\core\src\op\experimental_detectron_prior_grid_generator.cpp:104:
While validating node 'v6::ExperimentalDetectronPriorGridGenerator ExperimentalDetectronPriorGridGenerator_1766 (1069/Output_0/Data__const[0]:f32{3,4}, Conv_507[0]:f32{?,3,120,120}, 1070/Output_0/Data__const[0]:f32{1,1,1,1}) -> ()' with friendly_name 'ExperimentalDetectronPriorGridGenerator_1766':
The first dimension of both 'feature_map' and 'im_data' must match. Feature_map: ?; Im_data: 1
```

I also noticed that it works for `instance-segmentation-security-0228`. May be there is something specific about `instance-segmentation-security-1039`.

eaidova commented:
> The problem is still there, but it's not about demo.

This PR merging unblocks models validation, so we can see if issue is reproducible on the latest OV
